### PR TITLE
update tested python versions

### PR DESCRIPTION
**Pull request by davidism**
test 3.11 final
test 3.12 dev
test pypy 3.9
update config for tox 4

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
+    py3{12,11,10,9,8,7}
+    pypy39
     style
     typing
     docs
 skip_missing_interpreters = true
 
 [testenv]
+package = wheel
+wheel_build_env = .pkg
```
Comment (pgjones) [inline tox.ini:12]: Are these related to the versions?

File: tox.ini [hunk 2]
```diff
@@ -1,24 +1,31 @@
...
 deps = -r requirements/tests.txt
 commands = pytest -v --tb=short --basetemp={envtmpdir} {posargs}
 
 [testenv:style]
 deps = pre-commit
 skip_install = true
-commands = pre-commit run --all-files --show-diff-on-failure
```
Comment (pgjones) [inline tox.ini:16]: Is this intentional (not sure if it is useful, but I'd think it is)

File: tox.ini (hunk 1 above)
Comment (davidism) [reply tox.ini:12]: This is some new tox 4 config so that a single wheel is built for all test envs, rather than building and installing the sdist each time.

File: tox.ini (hunk 2 above)
Comment (davidism) [reply tox.ini:16]: Yeah, it's annoying when running locally (where you can use `git diff`), and workflow runs that would trigger this are already associated with a commit that shows a better diff.
